File: sky/engine/core/painting/RSTransform.dart
// Copyright 2015 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

part of dart_ui;

/// A transform consisting of a translation, a rotation, and a uniform scale.
///
/// Used by [Canvas.drawAtlas]. This is a more efficient way to represent these
/// simple transformations than a full matrix.
// Modeled after Skia's SkRSXform.
class RSTransform {
  /// Creates an RSTransform.
  ///
  /// An [RSTransform] expresses the combination of a translation, a rotation
  /// around a particular point, and a scale factor.
  ///
  /// The first argument, `scos`, is the cosine of the rotation, multiplied by
  /// the scale factor.
  ///
  /// The second argument, `ssin`, is the sine of the rotation, multiplied by
  /// that same scale factor.
  ///
  /// The third argument is the x coordinate of the translation, minus the
  /// `scos` argument multiplied by the x-coordinate of the rotation point, plus
  /// the `ssin` argument multiplied by the y-coordinate of the rotation point.
  ///
  /// The fourth argument is the y coordinate of the translation, minus the `ssin`
  /// argument multiplied by the x-coordinate of the rotation point, minus the
  /// `scos` argument multiplied by the y-coordinate of the rotation point.
  ///
  /// The [new RSTransform.fromComponents] method may be a simpler way to
  /// construct these values. However, if there is a way to factor out the
  /// computations of the sine and cosine of the rotation so that they can be
  /// reused over multiple calls to this constructor, it may be more efficient
  /// to directly use this constructor instead.
  RSTransform(double scos, double ssin, double tx, double ty) {
    _value
      ..[0] = scos
      ..[1] = ssin
      ..[2] = tx
      ..[3] = ty;
  }

  /// Creates an RSTransform from its individual components.
  ///
  /// The `rotation` parameter gives the rotation in radians.
  ///
  /// The `scale` parameter describes the uniform scale factor.
  ///
  /// The `anchorX` and `anchorY` parameters give the coordinate of the point
  /// around which to rotate.
  ///
  /// The `translateX` and `translateY` parameters give the coordinate of the
  /// offset by which to translate.
  ///
  /// This constructor computes the arguments of the [new RSTransform]
  /// constructor and then defers to that constructor to actually create the
  /// object. If many [RSTransform] objects are being created and there is a way
  /// to factor out the computations of the sine and cosine of the rotation
  /// (which are computed each time this constructor is called) and reuse them
  /// over multiple [RSTransform] objects, it may be more efficient to directly
  /// use the more direct [new RSTransform] constructor instead.
  factory RSTransform.fromComponents({
    double rotation,
    double scale,
    double anchorX,
    double anchorY,
    double translateX,
    double translateY
  }) {
    final double scos = math.cos(rotation) * scale;
    final double ssin = math.sin(rotation) * scale;
    final double tx = translateX + -scos * anchorX + ssin * anchorY;
    final double ty = translateY + -ssin * anchorX - scos * anchorY;
    return new RSTransform(scos, ssin, tx, ty);
  }

  final Float32List _value = new Float32List(4);

  /// The cosine of the rotation multiplied by the scale factor.
  double get scos => _value[0];

  /// The sine of the rotation multiplied by that same scale factor.
  double get ssin => _value[1];

  /// The x coordinate of the translation, minus [scos] multiplied by the
  /// x-coordinate of the rotation point, plus [ssin] multiplied by the
  /// y-coordinate of the rotation point.
  double get tx => _value[2];

  /// The y coordinate of the translation, minus [ssin] multiplied by the
  /// x-coordinate of the rotation point, minus [scos] multiplied by the
  /// y-coordinate of the rotation point.
  double get ty => _value[3];
}

File: engine/src/flutter/impeller/geometry/rstransform.cc
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "flutter/impeller/geometry/rstransform.h"

#include "flutter/impeller/geometry/matrix.h"

namespace impeller {

bool RSTransform::IsAxisAligned() const {
  return scaled_cos == 0.0f || scaled_sin == 0.0f;
}

Matrix RSTransform::GetMatrix() const {
  // clang-format off
  return Matrix::MakeRow(scaled_cos, -scaled_sin, 0.0f, translate_x,
                         scaled_sin,  scaled_cos, 0.0f, translate_y,
                            0.0f,       0.0f,     1.0f,     0.0f,
                            0.0f,       0.0f,     0.0f,     1.0f);
  // clang-format on
}

void RSTransform::GetQuad(Scalar width, Scalar height, Quad& quad) const {
  Point origin = {translate_x, translate_y};
  Point dx = width * Point{scaled_cos, scaled_sin};
  Point dy = height * Point{-scaled_sin, scaled_cos};
  quad = {
      // Ordered in the same Z pattern as Rect::GetPoints()
      origin,
      origin + dx,
      origin + dy,
      origin + dx + dy,
  };
}

Quad RSTransform::GetQuad(Scalar width, Scalar height) const {
  Quad quad;
  GetQuad(width, height, quad);
  return quad;
}

Quad RSTransform::GetQuad(Size size) const {
  Quad quad;
  GetQuad(size.width, size.height, quad);
  return quad;
}

std::optional<Rect> RSTransform::GetBounds(Scalar width, Scalar height) const {
  return Rect::MakePointBounds(GetQuad(width, height));
}

std::optional<Rect> RSTransform::GetBounds(Size size) const {
  return Rect::MakePointBounds(GetQuad(size));
}

}  // namespace impeller
